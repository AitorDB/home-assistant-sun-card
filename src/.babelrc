{
  "presets": [
    "@babel/env",
    "@babel/typescript"
  ],
  "plugins": [
    ["@babel/plugin-proposal-decorators", { "decoratorsBeforeExport": true }],
    "@babel/proposal-class-properties"
  ]
}
