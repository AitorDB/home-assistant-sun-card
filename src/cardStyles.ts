import { css } from 'lit-element'

export default css`
  .sun-card {
    --sun-card-lines: #464646;

    /* Use theme colors, fallback to default */
    --sun-card-text-color: var(--primary-text-color, #fff)
    --sun-card-subtitle-color: var(--secondary-text-color, #fff)

    color: var(--sun-card-text-color);
    padding: 1rem;
    
    font-family: var(--primary-font-family)
  }

  .sun-card-body {
    padding-top: 0.5rem;
  }

  .sun-card.sun-card-light {
    --sun-card-lines: #ececec;
    --sun-card-text-color: var(--primary-text-color, #000);
    --sun-card-subtitle-color: var(--secondary-text-color, #828282);
  }

  .sun-card-header {
    display: flex;
    justify-content: space-between;
  }
  
  .sun-card-footer .sun-card-footer-row {
    display: flex;
    justify-content: space-around;
    padding-top: 1.5rem;
  }

  .sun-card-title {
    font-size: 1.5rem;
    font-weight: 500;
    padding-bottom: 2rem;
    margin: 0;
  }

  .sun-card-text-container {
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  .sun-card-header .sun-card-text-subtitle {
    font-size: 1.15rem;
    font-weight: 400;
    padding-bottom: 0.25rem;
    color: var(--sun-card-subtitle-color);
  }

  .sun-card-header .sun-card-text-time {
    font-size: 1.85rem;
    font-weight: 400;
  }

  .sun-card-footer .sun-card-text-subtitle {
    font-size: 1.25rem;
    font-weight: 400;
    padding-bottom: 0.5rem;
    color: var(--sun-card-subtitle-color);
  }

  .sun-card-footer .sun-card-text-time {
    font-size: 1.25rem;
    font-weight: 500;
  }

  .sun-card-text-time-period {
    font-size: 0.75rem;
  }
`
