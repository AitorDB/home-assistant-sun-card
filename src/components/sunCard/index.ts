export * from './SunCard'
export * from './SunCardContent'
export * from './SunCardFooter'
export * from './SunCardGraph'
export * from './SunCardHeader'
