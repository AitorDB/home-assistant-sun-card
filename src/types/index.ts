import { LovelaceCardConfig } from 'custom-card-helpers'
import { I18N } from '../utils/I18N'


export type TSunCardFields = {
  sunrise?: boolean
  sunset?: boolean

  dawn?: boolean
  noon?: boolean
  dusk?: boolean

  azimuth?: boolean
  elevation?: boolean
}

export interface ISunCardConfig extends LovelaceCardConfig {
  i18n?: I18N
  darkMode?: boolean
  language?: string
  
  title?: string,
  component?: string
  use12hourClock?: boolean

  fields?: TSunCardFields
}

export interface IConfigChangedEvent <T> extends CustomEvent<T> {
  target: CustomEvent['target'] & {
    configValue: string
    selected?: string
    checked?: boolean
  }
}

export type TSunInfo = {
  sunrise: number,
  sunset: number

  dawnProgressPercent: number
  dayProgressPercent: number
  duskProgressPercent: number
  
  sunAboveHorizon: boolean
  sunPercentOverHorizon: number
  sunPosition: {
    x: number
    y: number
  }
}

export enum ESunCardErrors {
  SunIntegrationNotFound = 'SunIntegrationNotFound'
}

export type TSunCardData = {
  azimuth: number
  elevation: number

  sunInfo: TSunInfo

  times: TSunCardTimes

  error?: ESunCardErrors
}

export type TSunCardTimes = {
  dawn: Date
  dusk: Date
  noon: Date
  sunrise: Date
  sunset: Date
}

export enum ESunCardI18NKeys {
  Azimuth = 'azimuth',
  Dawn = 'dawn',
  Dusk = 'dusk',
  Elevation = 'elevation',
  Noon = 'noon',
  Sunrise = 'sunrise',
  Sunset = 'sunset'
}

export type TSunCardI18NKeys = { [key in ESunCardI18NKeys ]: string } | { errors: TSunCardI18NErrorKeys }

export type TSunCardI18N = Record<string, unknown>

export type TSunCardI18NErrorKeys = {
  [key in ESunCardErrors]: string
}
