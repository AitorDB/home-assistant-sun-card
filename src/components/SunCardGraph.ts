import { html, TemplateResult } from 'lit-html'
import { Constants } from '../constants'
import { TSunCardData, TSunInfo } from '../types'
export class SunCardGraph {

    private sunInfo: TSunInfo
  
    constructor (data: TSunCardData) {
      this.sunInfo = data?.sunInfo ?? Constants.DEFAULT_SUN_INFO
    }
  
    public render (): TemplateResult {
        const sunID = 'sun-gradient'
        const dawnID = 'dawn-gradient'
        const dayID = 'day-gradient'
        const duskID = 'dusk-gradient'

        const viewBox = "0 0 550 150"
    
        return html`
          <div class="sun-card-graph">
            <svg viewBox="${viewBox}" xmlns="http://www.w3.org/2000/svg">
              <defs>
                <linearGradient id="${sunID}" x1="0%" y1="0%" x2="0%" y2="100%">
                  <stop offset="0%" class="sunInitalStop" />
                  <stop offset="${this.sunInfo.sunPercentOverHorizon}%" class="sunMiddleStop" />
                  <stop offset="${this.sunInfo.sunPercentOverHorizon}%" class="sunEndStop" />
                </linearGradient>
                
                <linearGradient id="${dawnID}" x1="0%" y1="0%" x2="100%" y2="0%">
                  <stop offset="0%" class="dawnInitalStop" />
                  <stop offset="${this.sunInfo.dawnProgressPercent}%" class="dawnMiddleStop" />
                  <stop offset="${this.sunInfo.dawnProgressPercent}%" class="dawnEndStop" />
                </linearGradient>
                
                <linearGradient id="${dayID}" x1="0%" y1="0%" x2="100%" y2="0%">
                  <stop offset="0%" class="dayInitalStop" />
                  <stop offset="${this.sunInfo.dayProgressPercent}%" class="dayMiddleStop" />
                  <stop offset="${this.sunInfo.dayProgressPercent}%" class="dayEndStop" />
                </linearGradient>
                
                <linearGradient id="${duskID}" x1="0%" y1="0%" x2="100%" y2="0%">
                  <stop offset="0%" class="duskInitialStop" />
                  <stop offset="${this.sunInfo.duskProgressPercent}%" class="duskMiddleStop" />
                  <stop offset="${this.sunInfo.duskProgressPercent}%" class="duskEndStop" />
                </linearGradient>
              </defs>

              <path 
                class="sun-card-sun-line"
                d="M5,146 C29,153 73,128 101,108 C276,-29 342,23 449,108 C473,123 509,150 545,146"
                fill="none"
                stroke="var(--sun-card-lines)"
              />

              <path
                d="M5,146 C29,153 73,128 101,108 L 5 108"
                fill="url(#${dawnID})"
                stroke="url(#${dawnID})"
                opacity="${this.sunInfo.dawnProgressPercent}"
              />

              <path 
                d="M101,108 C276,-29 342,23 449,108 L 104,108"
                fill="url(#${dayID})"
                stroke="url(#${dayID})"
                opacity="${this.sunInfo.dayProgressPercent}"
              />

              <path 
                d="M449,108 C473,123 509,150 545,146 L 545 108"
                fill="url(#${duskID})"
                stroke="url(#${duskID})"
                opacity="${this.sunInfo.duskProgressPercent}"
              />

              <line x1="5" y1="108" x2="545" y2="108" stroke="var(--sun-card-lines)" />
              <line x1="101" y1="25" x2="101" y2="100" stroke="var(--sun-card-lines)" />
              <line x1="449" y1="25" x2="449" y2="100" stroke="var(--sun-card-lines)" />

              <circle
                cx="${this.sunInfo.sunPosition.x}"
                cy="${this.sunInfo.sunPosition.y}"
                r="17"
                opacity="${this.sunInfo.sunPercentOverHorizon}"
                stroke="none" fill="url(#${sunID})"
              />
            </svg>
          </div>
        `
    }
}
