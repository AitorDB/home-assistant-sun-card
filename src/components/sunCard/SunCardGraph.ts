import { html, TemplateResult } from 'lit-html'
import { Constants } from '../../constants'
import { TSunCardData, TSunInfo } from '../../types'
export class SunCardGraph {

    private sunInfo: TSunInfo
  
    constructor (data: TSunCardData) {
      this.sunInfo = data?.sunInfo ?? Constants.DEFAULT_SUN_INFO
    }
  
    public render (): TemplateResult {
        const sunID = 'sun-gradient'
        const dawnID = 'dawn-gradient'
        const dayID = 'day-gradient'
        const duskID = 'dusk-gradient'

        const viewBox = "0 0 550 150"
        // TODO: Check sun opacity
    
        return html`
          <div class="sun-card-graph">
            <svg viewBox="${viewBox}" xmlns="http://www.w3.org/2000/svg">
              
              <!-- Define gradients for use when drawing afterwards -->
              <defs>
                <linearGradient id="${sunID}" x1="0%" y1="0%" x2="0%" y2="100%">
                  <stop offset="0%" class="sunInitialStop" />
                  <stop offset="${this.sunInfo.sunPercentOverHorizon}%" class="sunMiddleStop" />
                  <stop offset="${this.sunInfo.sunPercentOverHorizon}%" class="sunEndStop" />
                </linearGradient>
                
                <linearGradient id="${dawnID}" x1="0%" y1="0%" x2="100%" y2="0%">
                  <stop offset="0%" class="dawnInitialStop" />
                  <stop offset="${this.sunInfo.dawnProgressPercent}%" class="dawnMiddleStop" />
                  <stop offset="${this.sunInfo.dawnProgressPercent}%" class="dawnEndStop" />
                </linearGradient>
                
                <linearGradient id="${dayID}" x1="0%" y1="0%" x2="100%" y2="0%">
                  <stop offset="0%" class="dayInitialStop" />
                  <stop offset="${this.sunInfo.dayProgressPercent}%" class="dayMiddleStop" />
                  <stop offset="${this.sunInfo.dayProgressPercent}%" class="dayEndStop" />
                </linearGradient>
                
                <linearGradient id="${duskID}" x1="0%" y1="0%" x2="100%" y2="0%">
                  <stop offset="0%" class="duskInitialStop" />
                  <stop offset="${this.sunInfo.duskProgressPercent}%" class="duskMiddleStop" />
                  <stop offset="${this.sunInfo.duskProgressPercent}%" class="duskEndStop" />
                </linearGradient>
              </defs>

              <!-- Draw path of the sun across the 'sky' -->
              <path 
                class="sun-card-sun-line"
                d="M5,146 C29,153 73,128 101,108 C276,-29 342,23 449,108 C473,123 509,150 545,146"
                fill="none"
                stroke="var(--sun-card-lines)"
              />

              <!-- Draw between the path of the sun and the horizon line for dawn -->
              <!-- IE: First dark blue part -->
              <path
                d="M5,146 C29,153 73,128 101,108 L 5 108"
                fill="url(#${dawnID})"
                stroke="url(#${dawnID})"
                opacity="${this.sunInfo.dawnProgressPercent}"
              />

              <!-- Draw between the path of the sun and the horizon line for day -->
              <!-- IE: Main light blue part in the middle -->
              <path 
                d="M101,108 C276,-29 342,23 449,108 L 104,108"
                fill="url(#${dayID})"
                stroke="url(#${dayID})"
                opacity="${this.sunInfo.dayProgressPercent}"
              />

              <!-- Draw between the path of the sun and the horizon line for dusk -->
              <!-- IE: Last dark blue part -->
              <path 
                d="M449,108 C473,123 509,150 545,146 L 545 108"
                fill="url(#${duskID})"
                stroke="url(#${duskID})"
                opacity="${this.sunInfo.duskProgressPercent}"
              />

              <!-- Draw the horizon, dawn and dusk lines (the gray horizontal/vertical lines) -->
              <line x1="5" y1="108" x2="545" y2="108" stroke="var(--sun-card-lines)" />
              <line x1="101" y1="25" x2="101" y2="100" stroke="var(--sun-card-lines)" />
              <line x1="449" y1="25" x2="449" y2="100" stroke="var(--sun-card-lines)" />

              <!-- Draw a circle representing the sun -->
              <circle
                cx="${this.sunInfo.sunPosition.x}"
                cy="${this.sunInfo.sunPosition.y}"
                r="17"
                opacity="${this.sunInfo.sunPercentOverHorizon}"
                stroke="none" fill="url(#${sunID})"
              />
            </svg>
          </div>
        `
    }
}
