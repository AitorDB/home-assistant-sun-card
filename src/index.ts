import { HomeAssistant } from "custom-card-helpers";
import { customElement, LitElement, state } from "lit-element";

import cardStyles from "./cardStyles";
import { Constants } from "./constants";
import { SunCardContent } from "./cardContent";
import { TSunCardConfig, TSunCardData, TSunCardTime } from "./types";

@customElement("sun-card")
class SunCard extends LitElement {
  static readonly cardType = "sun-card";
  static readonly cardName = "Sun Card";
  static readonly cardDescription =
    "Custom card that display a graph to track the sun position and related events";

  @state()
  private config: TSunCardConfig = {};

  @state()
  private data!: TSunCardData;

  private hasRendered = false;
  private lastHass!: HomeAssistant;

  set hass(hass: HomeAssistant) {
    this.lastHass = hass;

    if (!this.hasRendered) {
      return;
    }

    this.processLastHass();
  }

  calculatePositionAndProgressesByTime(hass: HomeAssistant) {
    const sunLine = this.shadowRoot?.querySelector("path") as SVGPathElement;
    const sunrise = new Date(hass.states["sun.sun"].attributes.next_rising);
    const sunset = new Date(hass.states["sun.sun"].attributes.next_setting);
    const eventsAt = {
      dayStart: 0,
      sunrise: this.convertDateToMinutesSinceDayStarted(sunrise),
      sunset: this.convertDateToMinutesSinceDayStarted(sunset),
      dayEnd: 23 * 60 + 59,
    };

    const now = new Date();
    const minutesSinceTodayStarted =
      this.convertDateToMinutesSinceDayStarted(now);

    // Dawn section position [0 - 105]
    const dawnSectionPosition =
      (Math.min(minutesSinceTodayStarted, eventsAt.sunrise) * 105) /
      eventsAt.sunrise;

    // Day section position [106 - 499]
    const minutesSinceDayStarted = Math.max(
      minutesSinceTodayStarted - eventsAt.sunrise,
      0
    );
    const daySectionPosition =
      (Math.min(minutesSinceDayStarted, eventsAt.sunset - eventsAt.sunrise) *
        (499 - 106)) /
      (eventsAt.sunset - eventsAt.sunrise);

    // Dusk section position [500 - 605]
    const minutesSinceDuskStarted = Math.max(
      minutesSinceTodayStarted - eventsAt.sunset,
      0
    );
    const duskSectionPosition =
      (minutesSinceDuskStarted * (605 - 500)) /
      (eventsAt.dayEnd - eventsAt.sunset);

    const position =
      dawnSectionPosition + daySectionPosition + duskSectionPosition;
    const sunPosition = sunLine.getPointAtLength(position);

    const dawnProgressPercent =
      (100 * (sunPosition.x - Constants.EVENT_X_POSITIONS.dayStart)) /
      (Constants.EVENT_X_POSITIONS.sunrise -
        Constants.EVENT_X_POSITIONS.dayStart);
    const dayProgressPercent =
      (100 * (sunPosition.x - Constants.EVENT_X_POSITIONS.sunrise)) /
      (Constants.EVENT_X_POSITIONS.sunset -
        Constants.EVENT_X_POSITIONS.sunrise);
    const duskProgressPercent =
      (100 * (sunPosition.x - Constants.EVENT_X_POSITIONS.sunset)) /
      (Constants.EVENT_X_POSITIONS.dayEnd - Constants.EVENT_X_POSITIONS.sunset);

    const sunYTop = sunPosition.y - Constants.SUN_RADIUS;
    const yOver = Constants.HORIZON_Y - sunYTop;
    let sunPercentOverHorizon = 0;
    if (yOver > 0) {
      sunPercentOverHorizon = Math.min(
        (100 * yOver) / (2 * Constants.SUN_RADIUS),
        100
      );
    }

    return {
      dawnProgressPercent,
      dayProgressPercent,
      duskProgressPercent,
      sunPercentOverHorizon,
      sunPosition: { x: sunPosition.x, y: sunPosition.y },
    };
  }

  convertDateToMinutesSinceDayStarted(date: Date) {
    return date.getHours() * 60 + date.getMinutes();
  }

  parseTime(timeText: string, locale?: string) {
    const regex = /\d{1,2}[:.]\d{1,2}|[AMP]+/g;
    const date = new Date(timeText);
    const { language, timeFormat } = this.getConfig();
    const result = date
      .toLocaleTimeString(locale ?? language, { hour12: timeFormat === "12h" })
      .match(regex) as [string, ("AM" | "PM")?];

    if (!result && !locale) {
      return this.parseTime(timeText, Constants.DEFAULT_CONFIG.language);
    }

    const [time, period] = result;
    return { time, period };
  }
  calculateTimeBetweenDates(date1: Date, date2: Date): TSunCardTime {
    const timeMs = Math.abs(date1 - date2);
    const { language } = this.getConfig();
    const time = new Date(timeMs).toLocaleTimeString(language, {
      hour12: false,
    });
    return { time };
  }
  processLastHass() {
    if (!this.lastHass) {
      return;
    }

    this.config.darkMode =
      this.config.darkMode ?? this.lastHass.themes.darkMode;
    this.config.language =
      this.config.language ??
      this.lastHass.locale?.language ??
      this.lastHass.language;
    this.config.timeFormat =
      this.config.timeFormat ??
      this.getTimeFormatByLanguage(this.config.language);

    const times = {
      dawn: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_dawn
      ),
      dusk: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_dusk
      ),
      noon: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_noon
      ),
      sunrise: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_rising
      ),
      sunset: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_setting
      ),
    };

    const {
      dawnProgressPercent,
      dayProgressPercent,
      duskProgressPercent,
      sunPercentOverHorizon,
      sunPosition,
    } = this.calculatePositionAndProgressesByTime(this.lastHass);

    const lengthOfDay = this.calculateTimeBetweenDates(
      new Date(this.lastHass.states["sun.sun"].attributes.next_rising),
      new Date(this.lastHass.states["sun.sun"].attributes.next_setting)
    );

    const daylightLeft = this.calculateTimeBetweenDates(
      new Date(this.lastHass.states["sun.sun"].attributes.next_setting),
      new Date()
    );

    const data: TSunCardData = {
      azimuth: this.lastHass.states["sun.sun"].attributes.azimuth,
      dawnProgressPercent,
      dayProgressPercent,
      duskProgressPercent,
      elevation: this.lastHass.states["sun.sun"].attributes.elevation,
      sunPercentOverHorizon,
      sunPosition,
      times,
      lengthOfDay,
      daylightLeft,
    };

    this.data = data;
  }

  getConfig() {
    const config: TSunCardConfig = {};
    config.darkMode = this.config.darkMode ?? Constants.DEFAULT_CONFIG.darkMode;
    config.language = this.config.language ?? Constants.DEFAULT_CONFIG.language;
    config.showAzimuth =
      this.config.showAzimuth ?? Constants.DEFAULT_CONFIG.showAzimuth;
    config.showElevation =
      this.config.showElevation ?? Constants.DEFAULT_CONFIG.showElevation;
    config.timeFormat =
      this.config.timeFormat ?? Constants.DEFAULT_CONFIG.timeFormat;
    config.title = this.config.title;

    if (
      !Object.keys(Constants.LOCALIZATION_LANGUAGES).includes(config.language!)
    ) {
      config.language = Constants.DEFAULT_CONFIG.language;
    }

    return config;
  }

  getTimeFormatByLanguage(language: string) {
    const date = new Date();
    const time = date.toLocaleTimeString(language).toLocaleLowerCase();
    return time.includes("pm") || time.includes("am") ? "12h" : "24h";
  }

  setConfig(config: TSunCardConfig) {
    this.config = { ...config };
  }

  protected render() {
    const config = this.getConfig();
    const language = config.language!;
    const localization = Constants.LOCALIZATION_LANGUAGES[language];
    return SunCardContent.generate(this.data, localization, config);
  }

  protected updated(changedProperties) {
    super.updated(changedProperties);

    if (!this.hasRendered) {
      this.hasRendered = true;
      this.processLastHass();
    }
  }

  static get styles() {
    return cardStyles;
  }
}

window.customCards = window.customCards || [];
window.customCards.push({
  type: SunCard.cardType,
  name: SunCard.cardName,
  description: SunCard.cardDescription,
});
