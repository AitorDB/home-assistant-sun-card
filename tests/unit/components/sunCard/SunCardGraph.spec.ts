import { SunCardGraph } from '../../../../src/components/sunCard/SunCardGraph'
import { TSunCardData } from '../../../../src/types'
import { CustomSnapshotSerializer, TemplateResultTestHelper } from '../../../helpers/TestHelpers'

expect.addSnapshotSerializer(new CustomSnapshotSerializer())

describe('SunCardGraph', () => {
  describe('render', () => {
    it(`should the graph with the data values when provided`, async () => {
      const data = {
        sunInfo: {
          dawnProgressPercent: 100,
          dayProgressPercent: 100,
          duskProgressPercent: 23,
          sunAboveHorizon: false,
          sunPercentOverHorizon: 0,
          sunPosition: {
            x: 50,
            y: 50
          },
          sunrise: 200,
          sunset: 400
        }
      } as TSunCardData

      const sunCardGraph = new SunCardGraph(data)
      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof sunCardGraph.render>
      element.templateResultFunction = () => sunCardGraph.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it(`should the graph with the default values when data values are not provided`, async () => {
      const data = {} as TSunCardData
      const sunCardGraph = new SunCardGraph(data)
      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof sunCardGraph.render>
      element.templateResultFunction = () => sunCardGraph.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })
  })
})
