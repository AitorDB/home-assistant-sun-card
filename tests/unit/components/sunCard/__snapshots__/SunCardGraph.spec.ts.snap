// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SunCardGraph render should the graph with the data values when provided 1`] = `
<!---->
      <div class="sun-card-graph">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 550 150">
          
          <!-- Define gradients for use when drawing afterwards -->
          <defs>
            <linearGradient x1="0%" y1="0%" x2="0%" y2="100%" id="sun-gradient">
              <stop offset="0%" class="sunInitialStop"></stop>
              <stop class="sunMiddleStop" offset="0%"></stop>
              <stop class="sunEndStop" offset="0%"></stop>
            </linearGradient>
            
            <linearGradient x1="0%" y1="0%" x2="100%" y2="0%" id="dawn-gradient">
              <stop offset="0%" class="dawnInitialStop"></stop>
              <stop class="dawnMiddleStop" offset="100%"></stop>
              <stop class="dawnEndStop" offset="100%"></stop>
            </linearGradient>
            
            <linearGradient x1="0%" y1="0%" x2="100%" y2="0%" id="day-gradient">
              <stop offset="0%" class="dayInitialStop"></stop>
              <stop class="dayMiddleStop" offset="100%"></stop>
              <stop class="dayEndStop" offset="100%"></stop>
            </linearGradient>
            
            <linearGradient x1="0%" y1="0%" x2="100%" y2="0%" id="dusk-gradient">
              <stop offset="0%" class="duskInitialStop"></stop>
              <stop class="duskMiddleStop" offset="23%"></stop>
              <stop class="duskEndStop" offset="23%"></stop>
            </linearGradient>
          </defs>

          <!-- Draw path of the sun across the 'sky' -->
          <path class="sun-card-sun-line" d="M5,146 C29,153 73,128 101,108 C276,-29 342,23 449,108 C473,123 509,150 545,146" fill="none" stroke="var(--sun-card-lines)"></path>

          <!-- Draw between the path of the sun and the horizon line for dawn -->
          <!-- IE: First dark blue part -->
          <path d="M5,146 C29,153 73,128 101,108 L 5 108" fill="url(#dawn-gradient)" stroke="url(#dawn-gradient)" opacity="100"></path>

          <!-- Draw between the path of the sun and the horizon line for day -->
          <!-- IE: Main light blue part in the middle -->
          <path d="M101,108 C276,-29 342,23 449,108 L 104,108" fill="url(#day-gradient)" stroke="url(#day-gradient)" opacity="100"></path>

          <!-- Draw between the path of the sun and the horizon line for dusk -->
          <!-- IE: Last dark blue part -->
          <path d="M449,108 C473,123 509,150 545,146 L 545 108" fill="url(#dusk-gradient)" stroke="url(#dusk-gradient)" opacity="23"></path>

          <!-- Draw the horizon, dawn and dusk lines (the gray horizontal/vertical lines) -->
          <line x1="5" y1="108" x2="545" y2="108" stroke="var(--sun-card-lines)"></line>
          <line x1="101" y1="25" x2="101" y2="100" stroke="var(--sun-card-lines)"></line>
          <line x1="449" y1="25" x2="449" y2="100" stroke="var(--sun-card-lines)"></line>

          <!-- Draw a circle representing the sun -->
          <circle r="17" stroke="none" cx="50" cy="50" opacity="0" fill="url(#sun-gradient)"></circle>
        </svg>
      </div>
    
`;

exports[`SunCardGraph render should the graph with the default values when data values are not provided 1`] = `
<!---->
      <div class="sun-card-graph">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 550 150">
          
          <!-- Define gradients for use when drawing afterwards -->
          <defs>
            <linearGradient x1="0%" y1="0%" x2="0%" y2="100%" id="sun-gradient">
              <stop offset="0%" class="sunInitialStop"></stop>
              <stop class="sunMiddleStop" offset="0%"></stop>
              <stop class="sunEndStop" offset="0%"></stop>
            </linearGradient>
            
            <linearGradient x1="0%" y1="0%" x2="100%" y2="0%" id="dawn-gradient">
              <stop offset="0%" class="dawnInitialStop"></stop>
              <stop class="dawnMiddleStop" offset="0%"></stop>
              <stop class="dawnEndStop" offset="0%"></stop>
            </linearGradient>
            
            <linearGradient x1="0%" y1="0%" x2="100%" y2="0%" id="day-gradient">
              <stop offset="0%" class="dayInitialStop"></stop>
              <stop class="dayMiddleStop" offset="0%"></stop>
              <stop class="dayEndStop" offset="0%"></stop>
            </linearGradient>
            
            <linearGradient x1="0%" y1="0%" x2="100%" y2="0%" id="dusk-gradient">
              <stop offset="0%" class="duskInitialStop"></stop>
              <stop class="duskMiddleStop" offset="0%"></stop>
              <stop class="duskEndStop" offset="0%"></stop>
            </linearGradient>
          </defs>

          <!-- Draw path of the sun across the 'sky' -->
          <path class="sun-card-sun-line" d="M5,146 C29,153 73,128 101,108 C276,-29 342,23 449,108 C473,123 509,150 545,146" fill="none" stroke="var(--sun-card-lines)"></path>

          <!-- Draw between the path of the sun and the horizon line for dawn -->
          <!-- IE: First dark blue part -->
          <path d="M5,146 C29,153 73,128 101,108 L 5 108" fill="url(#dawn-gradient)" stroke="url(#dawn-gradient)" opacity="0"></path>

          <!-- Draw between the path of the sun and the horizon line for day -->
          <!-- IE: Main light blue part in the middle -->
          <path d="M101,108 C276,-29 342,23 449,108 L 104,108" fill="url(#day-gradient)" stroke="url(#day-gradient)" opacity="0"></path>

          <!-- Draw between the path of the sun and the horizon line for dusk -->
          <!-- IE: Last dark blue part -->
          <path d="M449,108 C473,123 509,150 545,146 L 545 108" fill="url(#dusk-gradient)" stroke="url(#dusk-gradient)" opacity="0"></path>

          <!-- Draw the horizon, dawn and dusk lines (the gray horizontal/vertical lines) -->
          <line x1="5" y1="108" x2="545" y2="108" stroke="var(--sun-card-lines)"></line>
          <line x1="101" y1="25" x2="101" y2="100" stroke="var(--sun-card-lines)"></line>
          <line x1="449" y1="25" x2="449" y2="100" stroke="var(--sun-card-lines)"></line>

          <!-- Draw a circle representing the sun -->
          <circle r="17" stroke="none" cx="0" cy="0" opacity="0" fill="url(#sun-gradient)"></circle>
        </svg>
      </div>
    
`;
